core/issuers.chainmail

```diff
@@ -1,56 +1,174 @@
+ * mint.
+ */
+interface Mint {
+  /** Get the Issuer for this mint. */
+  getIssuer() -> (Issuer);
+
+  /**
+   * Create a new Purse containing the amount. Give it the specified name or "a
+   * purse".
+   */
+  mint(initialBalance:Amount, name:String = 'a purse') -> (Purse);
+
+  /** Reset all purses to contain nothing. */
+  destroyAll();
+
+  /** (Non-fungible amounts only.) Ensure no purse contains amount. */
+  destroy(amount:Amount);
+
+  /**
+   * Destroy any purses or payments containing amount, then issue a new Purse
+   * with that amount.
+   */
+  revoke(amount:Amount) -> (Purse);
+}
+
+/**
+ * Purses hold verified amounts of certain rights issued by Mints. Purses can
+ * transfer part of the balance they hold in a Payment, which has a narrower
+ * interface. A purse's balance can rise and fall, through the action of deposit
+ * and withdraw, while payments don't accept deposits.
+ *
+ * The primary use for Purses and Payments is for currency-like and goods-like
+ * valuables, but they can also be used to represent other kinds of rights, such
+ * as the right to participate in a particular contract.
+ *
+ * Currency-like assets are fungible, and don't distinguish use rights from
+ * transfer rights. The only use of a currency is to spend it. Some goods
+ * distinguish use rights from transfer rights, like a membership card; you can
+ * exercise the rights without giving them up.
+ */
+interface Purse extends (PurseOrPayment) {
+  /** Get the Issuer for this mint. */
+  getIssuer() -> (Issuer);
+  getXferBalance() -> (Amount);
+  getUseBalance() -> (Amount);
+  deposit(amount:Amount, srcPayment:PurseOrPayment) -> (Amount);
+  depositAll(srcPayment:PurseOrPayment) -> (Amount);
+  withdraw(amount:Amount, name:String) -> (Payment);
+  withdrawAll(name:String) -> (Payment);
 }
-makeNatAssay(label ::Label) ::Assay;
-makeMetaSingleAssayMaker(
-  baseLabelToAssayFn ::(Label -> Assay)) ::(Label -> Assay);
-
-class Issuer[Assay] {
-  type Amount = Assay.Amount;
-  getLabel() ::{ issuer ::Issuer, description };
-  getAssay() ::Assay;
-  makeEmptyPurse(name ?String) ::Purse;
-
-  class Mint {
-    getIssuer() Issuer;
-    mint(initialBalance ?Amount, name ?String) ::Purse;
-  }
-  class Payment {
-    getIssuer() ::Issuer;
-    getXferBalance() ::Amount;
-  }
-  class Purse {
-    getIssuer() ::Issuer;
-    getXferBalance() ::Amount;
-    getUseBalance() ::Amount;
-    deposit(amount ?Amount, srcPaymentP ?reveal[Promise]) ::Amount;
-    withdraw(amount ?Amount, name ?String) ::Purse;
-  }
+
+/**
+ * Payments hold verified amounts of certain rights issued by Mints. Amounts
+ * from payments can be deposited in purses, but otherwise, the entire amount is
+ * available when the payment is transferred. A payment's balance can only fall,
+ * through the action of withdraw.
+ * 
+ * Payments can be converted to Purses by calling
+ * payment.getIssuer().makeEmptyPurse().depositAll(payment);
```

Comment:
I like having this instruction, but getting the issuer from the payment when the payment may not be trusted is an anti-pattern, I think.

Reply:
Maybe the payment shouldn't even have a getIssuer()

Reply:
The issuer isn't intended to be tightly held. 

The new purse is as trustworthy as the payment, I believe.

[part of the reason I put that line here is as a call to create a shortcut.]

Reply:
I think we might be misunderstanding each other. The security mechanism for ensuring that I get a real payment is calling depositAll(payment) on a purse that I create from an issuer I trust. Getting the issuer from the payment doesn't provide any security, so it is an anti-pattern that we should not use. We need to get the issuer from other means for the check that depositAll() does to have meaning.

Reply:
I like this! It explains better what we're getting by separating purse from payment. A purse is what you hold onto to indicate a prior trust relationship. A payment is generally (though not always) something you're getting from someone else as an alleged payment. The APIs should be tuned towards this pattern of relationships.

I do not (yet?) conclude from this that we should remove `getIssuer` from payment. But it is the right question to ask.

Reply:
I'm currently feeling that getIssuer doesn't belong in payment. I re-wrote this hint in terms of a verified issuer.